File: surface/widgets/compose/components/widgets/data-display/DataTable.kt
// ============================================================
// Clef Surface Compose Widget — DataTable
//
// Sortable data table with configurable columns, row selection,
// and header sort indicators. Compose adaptation: LazyColumn
// with a sticky header Row for column labels, data Rows for
// each record, HorizontalDivider between rows, and optional
// Checkbox selection.
// See widget spec: repertoire/widgets/data-display/data-table.widget
// ============================================================

package clef.surface.compose.components.widgets.datadisplay

import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowDownward
import androidx.compose.material.icons.filled.ArrowUpward
import androidx.compose.material3.Checkbox
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

// --------------- Types ---------------

data class DataTableColumn(
    val key: String,
    val label: String,
    val width: Dp = 120.dp,
    val sortable: Boolean = false,
)

enum class SortDirection {
    Ascending,
    Descending,
    None,
}

// --------------- Component ---------------

/**
 * Sortable data table with configurable columns and optional row selection.
 *
 * @param columns Column definitions.
 * @param data Row data, each map keyed by column key.
 * @param sortColumn Currently sorted column key.
 * @param sortDirection Current sort direction.
 * @param selectable Whether rows are selectable via checkbox.
 * @param loading Whether data is loading.
 * @param emptyMessage Message shown when data is empty.
 * @param onSort Callback when a column header is tapped for sorting.
 * @param onSelectRow Callback when a row selection changes.
 * @param modifier Modifier for the root layout.
 */
@Composable
fun DataTable(
    columns: List<DataTableColumn>,
    data: List<Map<String, Any?>>,
    sortColumn: String? = null,
    sortDirection: SortDirection = SortDirection.None,
    selectable: Boolean = false,
    loading: Boolean = false,
    emptyMessage: String = "No data available",
    onSort: (column: String, direction: SortDirection) -> Unit = { _, _ -> },
    onSelectRow: (rowIndex: Int, selected: Boolean) -> Unit = { _, _ -> },
    modifier: Modifier = Modifier,
) {
    val selectedRows = remember { mutableStateListOf<Int>() }
    val scrollState = rememberScrollState()

    if (loading) {
        Box(
            modifier = modifier
                .fillMaxWidth()
                .padding(16.dp),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = "Loading...",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        return
    }

    if (data.isEmpty()) {
        Box(
            modifier = modifier
                .fillMaxWidth()
                .padding(16.dp),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = emptyMessage,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        return
    }

    LazyColumn(modifier = modifier.fillMaxWidth()) {
        // Header row
        item {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .horizontalScroll(scrollState)
                    .padding(vertical = 8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                if (selectable) {
                    Spacer(modifier = Modifier.width(48.dp))
                }
                columns.forEach { col ->
                    Row(
                        modifier = Modifier
                            .width(col.width)
                            .then(
                                if (col.sortable) {
                                    Modifier.clickable {
                                        val newDir = if (sortColumn == col.key && sortDirection == SortDirection.Ascending) {
                                            SortDirection.Descending
                                        } else {
                                            SortDirection.Ascending
                                        }
                                        onSort(col.key, newDir)
                                    }
                                } else {
                                    Modifier
                                },
                            ),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(
                            text = col.label,
                            style = MaterialTheme.typography.labelLarge,
                            fontWeight = FontWeight.Bold,
                        )
                        if (col.sortable && sortColumn == col.key) {
                            Icon(
                                imageVector = if (sortDirection == SortDirection.Ascending) {
                                    Icons.Filled.ArrowUpward
                                } else {
                                    Icons.Filled.ArrowDownward
                                },
                                contentDescription = sortDirection.name,
                                modifier = Modifier.padding(start = 4.dp),
                            )
                        }
                    }
                }
            }
            HorizontalDivider(thickness = 2.dp)
        }

        // Data rows
        itemsIndexed(data) { rowIndex, row ->
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .horizontalScroll(scrollState)
                    .padding(vertical = 8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                if (selectable) {
                    Checkbox(
                        checked = rowIndex in selectedRows,
                        onCheckedChange = { checked ->
                            if (checked) {
                                selectedRows.add(rowIndex)
                            } else {
                                selectedRows.remove(rowIndex)
                            }
                            onSelectRow(rowIndex, checked)
                        },
                        modifier = Modifier.width(48.dp),
                    )
                }
                columns.forEach { col ->
                    Text(
                        text = row[col.key]?.toString() ?: "",
                        style = MaterialTheme.typography.bodyMedium,
                        modifier = Modifier.width(col.width),
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }
            if (rowIndex < data.lastIndex) {
                HorizontalDivider()
            }
        }
    }
}

File: surface/widgets/swiftui/components/widgets/data-display/DataTable.swift
// ============================================================
// Clef Surface SwiftUI Widget — DataTable
//
// Sortable data table with configurable columns, row selection,
// and header sort indicators.
// ============================================================

import SwiftUI

// --------------- Types ---------------

struct DataTableColumn: Identifiable {
    let id: String
    let key: String
    let label: String
    var width: CGFloat = 120
    var sortable: Bool = false

    init(key: String, label: String, width: CGFloat = 120, sortable: Bool = false) {
        self.id = key
        self.key = key
        self.label = label
        self.width = width
        self.sortable = sortable
    }
}

enum SortDirection { case ascending, descending, none }

// --------------- Component ---------------

/// DataTable view with sortable columns and optional row selection.
///
/// - Parameters:
///   - columns: Column definitions.
///   - data: Row data as dictionaries keyed by column key.
///   - sortColumn: Currently sorted column key.
///   - sortDirection: Current sort direction.
///   - selectable: Whether rows are selectable.
///   - loading: Whether data is loading.
///   - emptyMessage: Message shown when data is empty.
///   - onSort: Callback when a column header is tapped.
///   - onSelectRow: Callback when a row selection changes.
struct DataTableView: View {
    var columns: [DataTableColumn]
    var data: [[String: String]]
    var sortColumn: String? = nil
    var sortDirection: SortDirection = .none
    var selectable: Bool = false
    var loading: Bool = false
    var emptyMessage: String = "No data available"
    var onSort: ((String, SortDirection) -> Void)? = nil
    var onSelectRow: ((Int, Bool) -> Void)? = nil

    @State private var selectedRows: Set<Int> = []

    var body: some View {
        if loading {
            HStack {
                Spacer()
                Text("Loading...")
                    .foregroundColor(.secondary)
                Spacer()
            }
            .padding(16)
        } else if data.isEmpty {
            HStack {
                Spacer()
                Text(emptyMessage)
                    .foregroundColor(.secondary)
                Spacer()
            }
            .padding(16)
        } else {
            ScrollView(.horizontal) {
                VStack(spacing: 0) {
                    // Header
                    HStack(spacing: 0) {
                        if selectable {
                            Spacer().frame(width: 48)
                        }
                        ForEach(columns) { col in
                            SwiftUI.Button(action: {
                                guard col.sortable else { return }
                                let newDir: SortDirection = (sortColumn == col.key && sortDirection == .ascending)
                                    ? .descending : .ascending
                                onSort?(col.key, newDir)
                            }) {
                                HStack(spacing: 4) {
                                    Text(col.label)
                                        .font(.subheadline)
                                        .fontWeight(.bold)
                                    if col.sortable && sortColumn == col.key {
                                        Image(systemName: sortDirection == .ascending ? "arrow.up" : "arrow.down")
                                            .font(.caption2)
                                    }
                                }
                            }
                            .buttonStyle(.plain)
                            .frame(width: col.width, alignment: .leading)
                        }
                    }
                    .padding(.vertical, 8)

                    Divider()

                    // Data rows
                    ForEach(Array(data.enumerated()), id: \.offset) { rowIndex, row in
                        HStack(spacing: 0) {
                            if selectable {
                                let isSelected = selectedRows.contains(rowIndex)
                                SwiftUI.Button(action: {
                                    if isSelected {
                                        selectedRows.remove(rowIndex)
                                    } else {
                                        selectedRows.insert(rowIndex)
                                    }
                                    onSelectRow?(rowIndex, !isSelected)
                                }) {
                                    Image(systemName: isSelected ? "checkmark.square.fill" : "square")
                                        .foregroundColor(.accentColor)
                                }
                                .buttonStyle(.plain)
                                .frame(width: 48)
                            }

                            ForEach(columns) { col in
                                Text(row[col.key] ?? "")
                                    .font(.body)
                                    .lineLimit(1)
                                    .frame(width: col.width, alignment: .leading)
                            }
                        }
                        .padding(.vertical, 8)

                        if rowIndex < data.count - 1 {
                            Divider()
                        }
                    }
                }
            }
        }
    }
}
